File: Day1-9/8.py
import itertools


def print_display(state):
    for y in range(len(state[0])):
        row = ''.join('#' if state[i][y] else '.' for i in range(len(state)))
        print(row)
    print('\n')


def process_cmd_list(cmds):
    # pixels[x][y]
    #   inner lists are columns
    pixels = [[False for i in range(6)] for j in range(50)]
    cmds = map(lambda x: x.strip().split(), cmds)
    for cmd in cmds:
        if cmd[0] == 'rect':
            x, y = map(int, cmd[1].split('x'))
            for i, j in itertools.product(range(x), range(y)):
                pixels[i][j] = True
        if cmd[0] == 'rotate':
            rc, loc, _, amt = cmd[1:]
            loc, amt = int(loc.split('=')[-1]), int(amt)
            # column rotation
            if rc == 'column':
                pixels[loc] = pixels[loc][-amt:] + pixels[loc][:-amt]
            # row rotation
            else:
                row = [pixels[i][loc] for i in range(len(pixels))]
                row = row[-amt:] + row[:-amt]
                for i in range(len(pixels)):
                    pixels[i][loc] = row[i]
    return sum(1 for i in pixels for j in i if j), pixels


if __name__ == '__main__':
    with open('8-test.txt', 'r') as f:
        assert process_cmd_list(f.readlines())[0] == 6
    with open('8.txt', 'r') as f:
        amt, res = process_cmd_list(f.readlines())
        print("Part 1: %s" % amt)
        print("Part 2:")
        print_display(res)

File: Day1-9/8.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define ROW_L 50
#define COL_L 6
#define SIZE 50 * 6

void print_display(char *disp){
    char out;
    for(int i = 0; i < SIZE; i++){
        if(i % 50 == 0 && i) printf("\n");
        out = disp[i] ? '#' : ' ';
        printf("%c", out);
    }
    printf("\n\n");
}

int sum_pixels(char *disp){
    int x = 0;
    for(int i = 0; i < SIZE; i ++) {
        if(disp[i]) x ++;
    }
    return x;
}

int posmod(int i, int n) {
    return (i % n + n) % n;
}

void decrypt(char* fname) {
    FILE *fp = fopen(fname, "r");
    char instr [100];
    char display [SIZE] = {0};
    while(fgets(instr, 100, fp) != NULL) {
        // Rect command
        if(strstr(instr, "rect")){
            char *num_seg = strstr(instr, " ") + 1;
            int x = atoi(strtok(num_seg, "x"));
            int y = atoi(strtok(NULL, "x"));
            // Fill in an x by y rectangle of pixels
            for(int i = 0; i < x; i++) {
                for(int j = 0; j < y; j++){
                    display[i + ROW_L * j] = 1;
                }
            }
        }
        // Rotate command
        else {
            char rc = strstr(instr, "column") == NULL;
            strtok(instr, "="); // Clear out before x/y=
            int loc = atoi(strtok(NULL, " "));
            strtok(NULL, " "); // Clear out " by "
            int amt = atoi(strtok(NULL, " "));

            // Row rotation
            if(rc) {
                char temp[ROW_L];
                for(int i = 0; i < ROW_L; i ++)
                    temp[i] = display[loc * ROW_L + posmod(i - amt, ROW_L)];
                for(int i = 0; i < ROW_L; i++)
                    display[loc * ROW_L + i] = temp[i];
            }
            // Column Rotation
            else {
                char temp[COL_L];
                for(int i = 0; i < COL_L; i ++)
                    temp[i] = display[loc + ROW_L * posmod(i - amt, COL_L)];
                for(int i = 0; i < COL_L; i ++)
                    display[loc + ROW_L * i] = temp[i];
            }
        }
    }
    printf("Part 1: %d\n", sum_pixels(display));
    printf("Part 2:\n");
    print_display(display);
}

int main() {
    decrypt("8.txt");
}